QPOptions (#5169)
knative/docs: docs/serving/queue-extensions.md

> # Queue Proxy Runtime Extensions

Reviewer: this title is not really descriptive; it doesn't tell a user why they would want to read this section or what information to expect

> Queue Proxy can be built with additional runtime extensions. Such extensions add optional features to Queue Proxy. While Knative core upstream introduce core features available on any knative deployment, Queue Proxy Extensions allow adding additional runtime features. It is up to Downstream to decide which such extensions will be available on what package.

Reviewer: doesn't explain what Queue Proxy is or why someone would want to extend it?

> Queue Proxy can be built with additional runtime extensions. Such extensions add optional features to Queue Proxy. While Knative core upstream introduce core features available on any knative deployment, Queue Proxy Extensions allow adding additional runtime features. It is up to Downstream to decide which such extensions will be available on what package.

Reviewer: ```suggestion
Queue Proxy can be built with additional runtime extensions enabled that provide optional features.
```
I'd simplify this by a lot and instead explain what Queue Proxy is and why it matters in Knative. The stuff about downstream etc should be explained by different vendors in their downstream docs instead, i.e. they can call out which additional features if any are available in their offering. For core Knative docs we should instead explain the feature and what the possible options are.

> For example, the [security-guard](https://knative.dev/security-guard) repository includes a Queue Proxy Extension [qpoption](https://knative.dev/security-guard/pkg/qpoption) that offers the ability to add security plus to Queue Proxy. Other extensions may be developed to add other features.

Reviewer: ```suggestion
For example, the [`security-guard`](https://knative.dev/security-guard) repository provides a Queue Proxy Option [qpoption](https://knative.dev/security-guard/pkg/qpoption) package, which enables users to add additional security features to Queue Proxy.
```

> The runtime features available are determined when the Queue Proxy image is built. Queue Proxy defines an orderly manner to activate and to configure extensions - See:
> 
> - [Enabling Queue Proxy Pod Info](./configuration/feature-flags.md#queue-proxy-pod-info) - discussing a necessary step to enable the use of Extensions.
> - [Using Queue Proxy Extensions](./services/using-queue-extensions.md) - discussing how to configure a service to use features implemented in extensions.

Reviewer: ```suggestion
The runtime features available are determined when the Queue Proxy image is built. Queue Proxy defines an orderly manner to activate and to configure extensions.

#### Additional information

- [Enabling Queue Proxy Pod Info](./configuration/feature-flags.md#queue-proxy-pod-info) - discussing a necessary step to enable the use of Extensions.
- [Using Queue Proxy Extensions](./services/using-queue-extensions.md) - discussing how to configure a service to use features implemented in extensions.
```

> In order to add extensions, `cmd/queue/main.go` need to be replaced before queue proxy image is built. Here is an example `cmd/queue/main.go` that adds the `test-gate` extension.

Reviewer: ```suggestion
#### Adding extentions

You can add extensions by replacing the `cmd/queue/main.go` file before the Queue Proxy image is built. The following example shows a `cmd/queue/main.go` file that adds the `test-gate` extension:
```

# Extending Queue Proxy image with QPoptions

Knative service pods include two containers:

- The user main service container, which is named `user-container`
- The Queue Proxy - a sidecar named `queue-proxy` that serves as a reverse proxy in front of the `user-container`

> Queue Proxy can be extended to offer additional optional features. The QPOptions feature of Queue Proxy allows additional runtime packages to extend Queue Proxy capabilities.

Reviewer 2: ```suggestion
You can extend Queue Proxy to offer additional features. The QPOptions feature of Queue Proxy allows additional runtime packages to extend Queue Proxy capabilities.
```

Reviewer 2: Should it be "QPOptions" or "QPoptions" here? Make sure one or the other is consistently used.

> For example, the [security-guard](https://knative.dev/security-guard/README.md) repository provides an extension that uses the QPOptions feature. The [Qpoption](https://knative.dev/security-guard/pkg/qpoption/README.md) package enables users to add additional security features to Queue Proxy.

Reviewer 2: Check "QPOptions" vs "QPoptions" here also.

Author: will do a full search and align in a separate commit

The runtime features available are determined when the Queue Proxy image is built. Queue Proxy defines an orderly manner to activate and to configure extensions.

> #### Additional information

Reviewer 2: ```suggestion
## Additional information
```

Reviewer 2: Don't use an H4 heading that's not preceded by an H3.

> - [Enabling Queue Proxy Pod Info](./configuration/feature-flags.md#queue-proxy-pod-info) - discussing a necessary step to enable the use of Extensions.

Reviewer 2: ```suggestion
- [Enabling Queue Proxy Pod Info](./configuration/feature-flags.md#queue-proxy-pod-info) - discussing a necessary step to enable the use of extensions.
```

- [How to use extensions enabled by QPoptions](./services/using-queue-extensions.md) - discussing how to configure a service to use features implemented in extensions.

> #### Adding extensions

Reviewer 2: ```suggestion
## Adding extensions
```
